### PR TITLE
OnehotCategoricalProjection

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
This PR implements the ``OnehotCategoricalProjectionNetwork`` that provides option to use the `straight-through` gradient estimator.

Straight through gradient estimator is proposed by Bengio et al., "Estimating or Propagating Gradients Through Stochastic Neurons for Conditional Computation", 2013.

This is used in some RL algorithms such as Dreamer.

### CODE REVIEW

File: alf/networks/projection_networks.py [hunk 1]
```diff
@@ -765,3 +765,68 @@ def forward(self, inputs, state=()):
             upper_bound=self._action_high)
 
         return dist, state
+
+
+@alf.configurable
+class OnehotCategoricalProjectionNetwork(Network):
```
Comment (hnyu) [inline alf/networks/projection_networks.py:771]: This seems more appropriately a transformation/wrapper on top of the existing CategoricalProjectionNetwork's output dist: 
``` python
dist = CategoricalProjectionNetwork(...)
dist1 = OnehotCategoricalTransform(dist)
```

Comment (Haichao-Zhang) [reply alf/networks/projection_networks.py:771]: > This seems more appropriately a transformation/wrapper on top of the existing CategoricalProjectionNetwork:
> dist = OnehotCategorical(Categorical)
> 

The suggested one is a possible alternative approach, but my current thought is that it is unnecessary as it creates another layer of complication for the user.

File: alf/utils/dist_utils.py [hunk 2]
```diff
@@ -560,6 +560,23 @@ def _get_categorical_builder(obj: td.Categorical):
...
+        return td.OneHotCategorical, {'logits': obj.logits}
+
+
+def _get_onehot_categorical_st_builder(
+        obj: td.OneHotCategoricalStraightThrough):
+    if 'probs' in obj.__dict__ and id(obj.probs) == id(obj._param):
+        # This means that obj is constructed using probs
+        return td.OneHotCategoricalStraightThrough, {'probs': obj.probs}
+    else:
+        return td.OneHotCategoricalStraightThrough, {'logits': obj.logits}
```
Comment (hnyu) [inline alf/utils/dist_utils.py]: You can probably merge the three functions above.

Comment (Haichao-Zhang) [reply alf/utils/dist_utils.py]: Good point. Merged.

**Review by hnyu — APPROVED** [review]
